### PR TITLE
Improve diagnostics for optionals in string interpolation segments

### PR DESCRIPTION
**Pull request by CodaFi**
Improvements to the diagnostics from #5110 as suggested by @jrose-apple and @jtbandes.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by jtbandes** [comment]
I guess what I meant about "make this explicit" is that it doesn't seem very clear what "this" is from the beginning of the sentence. This is making me think we should have some sort of style guide for diagnostics :)

**Comment by CodaFi** [comment]
`this` refers to `a debug description for an optional value`.  Do you have a better aside that motivates the fixits?

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by jtbandes** [comment]
Looking at some other diagnostics, I think it's somewhat common for the initial warning to *not* provide extra motivation that's evident from the fixits (which are considered part of the diagnostic output).

For example, just above your test, there's `warning: expression implicitly coerced from 'Int?' to Any` / `note: provide a default value to avoid this warning`. IMO it's sufficiently clear when the warning just describes what's wrong and the fixits give you multiple options for how to fix it. (Without doing extensive research, I think I've more commonly seen motivating text like `…; did you mean ___?` when there's only *one* fixit given.)

**Comment by CodaFi** [comment]
Look through diagnostics for Sema and you'll find about 50 with independent clauses justifying or clarifying some part or another.  [I'm pushing back on this because we tried being terse in the initial presentation and agreed it was vague.]

**Comment by CodaFi** [comment]
@swift-ci please smoke test linux platform.

**Comment by jrose-apple** [comment]
@jtbandes: There's a difference between fix-its on notes and fix-its on warnings: the latter is supposed to represent the right choice 90% of the time, whereas a note is just a suggestion. In the warning case, the text has to include a description of both the problem and the fix, but when it's just a note usually it just describes the fix, possibly along with the interpretation of the code that leads to the fix.

**Comment by jrose-apple** [comment]
@CodaFi I thought we agreed to eliminate one or the other of the fix-its? I know they're not 100% equivalent at this point but it still seems like overkill. And I still think it's worth having the {{?? <#default#>}} fix-it as well; was there more discussion there to be had?

**Comment by CodaFi** [comment]
Joe pointed out `debugDescription` being an inappropriate fix, but I don't remember suggesting *less* fixits being on the table.  Everybody seemed to want a different set of fixits for this on the list, and the one that @harlanhaskins and I agreed should stay is one that suggests explicit conversion to a reflected String.  All others are helpful but secondary and we have to balance fixing the problem with offering too many fixes for it.

**Comment by CodaFi** [comment]
All the methods in the fixits for Optional-to-Any conversion apply here - and at one point I had them factored out as much.  I'm just not sure offering 5 fixits is a good thing to do here.

**Comment by jrose-apple** [comment]
We don't need 5 fix-its, but I think we do need 2: "print as an Optional" and "provide a default". Right now both `String(describing:)` and `as Optional` result in the first behavior and there's no fix-it for the second.

(I guess the only person objecting was me, https://github.com/apple/swift/pull/5110#issuecomment-252117909)

**Comment by harlanhaskins** [comment]
@jrose-apple I agree with that. Ideally there are two fixes, one which silences the warning if you intended the behavior and one which fixes the problem if you didn't. Having both notes silence the warning IMO betrays the spirit of the warning. I'm leaning toward String(describing:) and the default as the 2 fix-its.

**Comment by harlanhaskins** [comment]
I'd say, though, that in most cases the _real_ fix is hoisting the optional into an if-let with an else clause that explicitly handles the `.none` case 😅

**Comment by CodaFi** [comment]
Then, to be absolutely clear, we want the following two fixits:

1) `String(describing:)`
2) `?? <default value>`

Agreed?

**Comment by jrose-apple** [comment]
@rudkx, additional thoughts?

**Comment by rudkx** [comment]
I'm not sure if nil-coalescing is really a better choice than just suggesting force-unwrapping (and letting the user use their judgement of whether that makes sense, or whether nil-coalescing or `if-let` makes more sense for their particular case).

I don't feel strongly either way, though, so do what you think makes the most sense!

**Comment by CodaFi** [comment]
@swift-ci please smoke test.

**Comment by CodaFi** [comment]
Swift Package Manager failures look strange.  @shahmishal Does this require a clean test?

**Comment by shahmishal** [comment]
@swift-ci please smoke test Linux

**Comment by CodaFi** [comment]
@swift-ci please clean smoke test linux platform.

**Comment by CodaFi** [comment]
⛵️

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -3723,28 +3723,26 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
...
             .highlight(segment->getSourceRange())
-            .fixItInsert(segment->getEndLoc(), ".debugDescription");
+            .fixItInsert(segmentStart, "String(describing: ")
+            .fixItInsert(segment->getEndLoc(), ")");
 
-          auto opts = PrintOptions::printForDiagnostics();
-          TC.diagnose(segment->getLoc(),
-                      diag::silence_optional_in_interpolation_segment_cast,
-                      segmentTy)
+          // Suggest a cast to 'Optional'.
```
Comment (jrose-apple) [inline lib/Sema/MiscDiagnostics.cpp]: Comment didn't get updated. :-/

Comment (CodaFi) [inline lib/Sema/MiscDiagnostics.cpp]: 😳.  Guess I'all have to fix something in the type checker to get to this...

**Pull request merged**